include/drivers/lora.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019 Manivannan Sadhasivam
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_LORA_H_
+#define ZEPHYR_INCLUDE_DRIVERS_LORA_H_
+
+/**
+ * @file
+ * @brief Public LoRa APIs
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+enum lora_signal_bandwidth {
+	BW_125_KHZ = 0,
+	BW_250_KHZ,
+	BW_500_KHZ,
+};
+
+enum lora_datarate {
+	SF_6 = 6,
+	SF_7,
+	SF_8,
+	SF_9,
+	SF_10,
+	SF_11,
+	SF_12,
+};
+
+enum lora_coding_rate {
+	CR_4_5 = 1,
+	CR_4_6 = 2,
+	CR_4_7 = 3,
+	CR_4_8 = 4,
+};
+
+struct lora_modem_config {
+	u32_t frequency;
+	enum lora_signal_bandwidth bandwidth;
+	enum lora_datarate datarate;
+	enum lora_coding_rate coding_rate;
+	u16_t preamble_len;
+	s8_t tx_power;
+	bool tx;
+};
+
+/**
+ * @typedef lora_api_config()
+ * @brief Callback API for configuring the LoRa module
+ *
+ * @see lora_config() for argument descriptions.
+ */
+typedef int (*lora_api_config)(struct device *dev,
+			       struct lora_modem_config *config);
+
+/**
+ * @typedef lora_api_send()
+ * @brief Callback API for sending data over LoRa
+ *
+ * @see lora_send() for argument descriptions.
+ */
+typedef int (*lora_api_send)(struct device *dev,
+			     u8_t *data, u32_t data_len);
+
+/**
+ * @typedef lora_api_recv()
+ * @brief Callback API for receiving data over LoRa
+ *
+ * @see lora_recv() for argument descriptions.
+ */
+typedef int (*lora_api_recv)(struct device *dev, u8_t *data, s32_t timeout);
+
+struct lora_driver_api {
+	lora_api_config config;
+	lora_api_send	send;
+	lora_api_recv	recv;
+};
+
+/**
+ * @brief Configure the LoRa modem
+ *
+ * @param dev     LoRa device
+ * @param config  Data structure containing the intended configuration for the
+		  modem
+ * @return 0 on success, negative on error
+ */
+static inline int lora_config(struct device *dev,
+			      struct lora_modem_config *config)
+{
+	const struct lora_driver_api *api = dev->driver_api;
+
+	return api->config(dev, config);
+}
+
+/**
+ * @brief Send data over LoRa
+ *
+ * @note This is a non-blocking call.
+ *
+ * @param dev       LoRa device
+ * @param data      Data to be sent
+ * @param data_len  Length of the data to be sent
+ * @return 0 on success, negative on error
+ */
+static inline int lora_send(struct device *dev,
+			    u8_t *data, u32_t data_len)
+{
+	const struct lora_driver_api *api = dev->driver_api;
+
+	return api->send(dev, data, data_len);
+}
+
+/**
+ * @brief Receive data over LoRa
+ *
+ * @note This is a blocking call.
+ *
+ * @param dev       LoRa device
+ * @param data      Buffer to hold received data
+ * @param timeout   Timeout value in milliseconds. API also accepts, K_NO_WAIT
+		    for no wait time and K_FOREVER for blocking until
+		    data arrives.
+ * @return Length of the data received on success, negative on error
+ */
+static inline int lora_recv(struct device *dev, u8_t *data, s32_t timeout)
```

Comment:
This API is missing a parameter that would indicate the size of the `data` buffer. How would the implementation know how many bytes can it copy into the `data` buffer? If we blindly copy data w/o taking buffer size into consideration we will sooner or later (but rather sooner) overwrite some data.

Reply:
I agree. I considered this while developing the API but I don't remember the reason why I decided to leave the size parameter :P

Reply:
@rlubos Okay, I got the reason now. The buffer copy happens with the size of the recevied length reported by the lora stack. So even if we pass the size parameter, we can't just copy those bytes since it may or may not be received. But there are couple of usecases with size parameter:

* We can check if the buffer size exceeded the max recv length i.e., 255
* Return error if we don't received the `size` bytes

Reply:
@Mani-Sadhasivam The main reason for providing an extra `size` parameter is to detect when we can't copy the data so that we don't exceed the buffer provided. I'm not a Lora expert, so I'm not sure what would be the best action to take in case this happens, but the solutions I see:
* Return an error and drop the packet (results in data loss),
* Copy only the bytes that fit, return a number of bytes copied and drop the rest (that's how DATAGRAM sockets work),
* Copy only the bytes that fit, return a number of bytes copied and keep the rest of the packet for further reads (that's how STREAM packet works)

I understand though, that any of these will require some rework and would postpone this PR. Alternatively, we could specify in the API what's the size of the buffer that needs to be provided (as according to @alexanderwachter I understand that Lora packet has a maximum size, if it's 255 bytes used in the sample I guess it could also be a reasonable approach).